### ISSUE TITLE
autocompletion for variable in GDScript

### ISSUE THREAD

**Issue by survivant** [godotengine/godot#15718]
I connected an Area2D to a kineticBody2D and in the script I have a method like that

```
func _on_Area2D_body_exited( body ):
	body.
```
when I enter body.   I don't hate the auto completing available.

GDEditor should be able to recognize the API and give us the method available for body.

**Comment by survivant** [comment]
I see.  We could have autocomplete like that.  you use the lower interface by default, in this case : PhysicsBody2D  and could give separate method by implementation.

ex : 
body.a   (list the method in PhysicsBody2D first, and the methods in StaticBody2D (if any)...

ex : 

body.a  should this

body.add_collision_exception_with(...)    (PhysicsBody2D )
body.add_force(...)   (RigidBody2D)

?
the goal is just to make it easier to find methods available

PS (or could it be possible to have different implementations ?  )
like
````
func _on_Area2D_body_exited( StaticBody2D body ):
	body.
func _on_Area2D_body_exited( RigidBody2D body ):
	body.
func _on_Area2D_body_exited( KinematicBody2D body ):
	body.

````

**Comment by manokara** [comment]
GDScript is dynamically typed, although there are plans for adding type hints to the language. Your first suggestion is plausible, but then again it's dependent on the implementation of type hinting **and** the possibility of adding a type (or more) for a parameter in the signal's declaration, which is complicated given that signal callbacks don't have fixed parameters, they are called with an array of values. Given the language's nature, method overloading doesn't work: you either have one type or you have any. Supporting all that would require a massive change to the language.

If you want to deal with different types of bodies, you use the `is` keyword (`extends` in Godot 2.x):

    func _on_Area2D_body_exited(body):
        if body is StaticBody2D:
            body.

Then, perhaps, the editor could analyse if the variable was checked with `is` previously and suggest the appropriate completions. But is that really feasible?

**Comment by akien-mga** [comment]
> Then, perhaps, the editor could analyse if the variable was checked with `is` previously and suggest the appropriate completions. But is that really feasible?

Actually it does exactly that :)

**Comment by manokara** [comment]
Oh, I never realized that 🤔  That's great!

**Comment by hilfazer** [comment]
Maybe assert() could enable autocompletion as well?

`assert( body is StaticBody2D)`

**Comment by reduz** [comment]
it could be possible to guess by checking where the signal originated...
however, code completion will work if you write:

```
if (body is KinematicBody):
    here completion will work, as it will understand the type

eventually in 3.1 the idea is to add optional typing, so you can write the type

**Comment by eon-s** [comment]
Once optional typing is added, some signals can have typed parameters by default (PhysicsBody2D in this case).

**Comment by akien-mga** [comment]
This is now fixed in 3.1 with this optional type hint:
```
func _on_Area2D_body_exited(body : PhysicsBody2D):
	body.pos<suggests "position">
``

**Comment by akien-mga** [comment]
BTW this was a duplicate of #5672.
